needle-slack-connector/src/server/backend/connectors-backend.ts

```diff
@@ -0,0 +1,314 @@
+    }
+  }
+
+  // Compute canvas differences excluding deleted channels
+  const canvasDiff = computeCanvasDiff(
+    canvasFiles
+      .filter((file) =>
+        existingChannelIds.has((file.metadata as CanvasFileMetadata).channelId),
+      )
+      .map((file) => ({
+        ndlFileId: file.ndlFileId,
+        metadata: file.metadata as CanvasFileMetadata,
+        updatedAt: file.updatedAt,
+      })),
+    liveCanvases.map((canvas) => ({
+      channelId: canvas.channelId,
+      originId: canvas.originId,
+      url: canvas.url,
+      title: canvas.title,
+      createdAt: canvas.createdAt,
+      updatedAt: canvas.updatedAt,
+      dataType: "canvas" as const,
+    })),
+  );
+
+  const canvasFileMap = new Map(
+    canvasFiles.map((file) => [
+      `${(file.metadata as CanvasFileMetadata).channelId}-${(file.metadata as CanvasFileMetadata).originId}`,
+      file.ndlFileId,
+    ]),
+  );
+
+  // Prepare canvas operations
+  const [canvasFilesToCreate, canvasFilesToUpdate, canvasFilesToDelete] = [
+    canvasDiff.create.map((canvas) => ({
+      id: createNeedleFileId(),
+      metadata: {
+        channelId: canvas.channelId,
+        dataType: "canvas" as const,
+        originId: canvas.originId,
+        title: canvas.title,
+        url: canvas.url,
+      },
+      title: canvas.title,
+    })),
+    canvasDiff.update.map((canvas) => ({
+      id: canvasFileMap.get(`${canvas.channelId}-${canvas.originId}`)!,
+      metadata: {
+        channelId: canvas.channelId,
+        dataType: "canvas" as const,
+        originId: canvas.originId,
+        title: canvas.title,
+        url: canvas.url,
+      },
+      title: canvas.title,
+    })),
+    [
+      ...canvasDiff.delete.map((canvas) => ({
+        id: canvasFileMap.get(`${canvas.channelId}-${canvas.originId}`)!,
+      })),
+      ...deletedChannelFiles.map((file) => ({ id: file.ndlFileId })),
+    ],
+  ];
+
+  // Merge message and canvas operations into a single descriptor
+  const descriptor: ConnectorRunDescriptor = {
+    create: [
+      ...create,
+      ...canvasFilesToCreate.map((canvas) => ({
+        id: canvas.id,
+        url: canvas.metadata.url,
+        type: "text/html" as const,
+      })),
+    ],
+    update: [
+      ...update,
+      ...canvasFilesToUpdate.map((canvas) => ({ id: canvas.id })),
+    ],
+    delete: [
+      ...deleteFiles,
+      ...canvasFilesToDelete.map((canvas) => ({ id: canvas.id })),
+    ],
+  };
+
+  await publishConnectorRun(connectorId, descriptor);
+
+  await handleDatabaseUpdates(
+    connectorId,
+    [...filesToCreate, ...canvasFilesToCreate],
+    [...filesToUpdate, ...canvasFilesToUpdate],
+    [...filesToDelete, ...canvasFilesToDelete],
+  );
+
+  await updateLastSynced(connectorId);
+
+  // Update channelInfo in the database if needed
+  await db
+    .update(slackConnectorsTable)
+    .set({ channelInfo: connectorDetails.channelInfo })
+    .where(eq(slackConnectorsTable.connectorId, connectorId));
+}
```

Comment:
wow monster function 😄 would be great if we can divide it into a few pieces.

Reply:
I will refactor, indeed pretty chaotic